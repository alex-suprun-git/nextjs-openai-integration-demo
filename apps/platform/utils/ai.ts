import { getLocale } from 'next-intl/server';
import { ChatOpenAI, OpenAIEmbeddings } from '@langchain/openai';
import { OutputFixingParser, StructuredOutputParser } from 'langchain/output_parsers';
import { PromptTemplate } from '@langchain/core/prompts';
import { Document } from 'langchain/document';
import { loadQARefineChain } from 'langchain/chains';
import { MemoryVectorStore } from 'langchain/vectorstores/memory';
import z from 'zod';

const modelParams = {
  model: 'o4-mini',
};

const parser = StructuredOutputParser.fromZodSchema(
  z.object({
    mood: z.string().describe('the mood of the person who wrote the journal entry.'),
    subject: z.string().describe('the subject of the journal entry.'),
    negative: z
      .boolean()
      .describe('is the journal entry negative? (i.e. does it contain negative emotions?).'),
    summary: z.string().max(100).describe('quick summary of the entire entry.'),
    color: z
      .string()
      .describe(
        'a hexidecimal color code that represents the mood of the entry. Example #0101fe for blue representing happiness.',
      ),
    sentimentScore: z
      .number()
      .describe(
        'sentiment of the text and rated on a scale from -10 to 10, where -10 is extremely negative, 0 is neutral, and 10 is extremely positive.',
      ),
    title: z.string().describe('A concise title summarizing the journal entry.'),
  }) as any,
);

const getPrompt = async (content: string, language: UserLocale) => {
  const formatInstructions = parser.getFormatInstructions();

  const templates = {
    de: 'Analysiere den folgenden Tagebucheintrag und antworte auf Deutsch. Übersetze die Formatierungsanweisungen ins Deutsche und befolge sie unbedingt! \n{formatInstructions}\n{entry}',
    en: 'Analyze the following journal entry and respond in English. Follow the instructions and format your response to match the format instructions, no matter what! \n{formatInstructions}\n{entry}',
  };

  const prompt = new PromptTemplate({
    template: templates[language],
    inputVariables: ['entry'],
    partialVariables: { formatInstructions },
  });

  const input = await prompt.format({
    entry: content,
  });

  return input;
};

export const analyzeEntry = async (content: string) => {
  const locale = await getLocale();

  const input = await getPrompt(content, locale as UserLocale);
  const model = new ChatOpenAI({
    ...modelParams,
    apiKey: process.env.OPENAI_API_KEY,
  });
  const output = await model.invoke(input);

  try {
    return await parser.parse(output.content as string);
  } catch (e) {
    const fixParser = OutputFixingParser.fromLLM(new ChatOpenAI(modelParams), parser);
    const fix = await fixParser.parse(output.content as string);
    return fix;
  }
};

export const analysisFeedback = async (
  language: UserLocale,
  question: string,
  entries: BaseEntry[],
) => {
  if (!entries.length) {
    return undefined;
  }

  const docs = entries.map(
    (entry) =>
      new Document({
        pageContent: entry.content as string,
        metadata: { source: entry.id, date: entry.createdAt },
      }),
  );
  const model = new ChatOpenAI(modelParams);
  const chain = loadQARefineChain(model);
  const embeddings = new OpenAIEmbeddings();
  const store = await MemoryVectorStore.fromDocuments(docs, embeddings);
  const relevantDocs = await store.similaritySearch(question);

  const templates = {
    en: `Answer the following question focusing exclusively on the person's mood and feelings based on the given documents. If the question touches on any other topic or if the context is unclear, please respond with a polite request to rephrase the question. Ensure your answer is in the same language as the input. \nQuestion: {question}`,
    de: `Beantworte die folgende Frage ausschließlich im Hinblick auf die Stimmung und Gefühle der Person basierend auf den gegebenen Dokumenten. Falls die Frage ein anderes Thema betrifft oder der Kontext unklar ist, bitte höflich um eine Umformulierung der Frage. Stelle sicher, dass deine Antwort in derselben Sprache wie die Eingabe erfolgt. \nFrage: {question}`,
  };

  const promptTemplate = new PromptTemplate({
    template: templates[language],
    inputVariables: ['question'],
  });

  const input = await promptTemplate.format({
    question: question,
  });

  const res = await chain.invoke({
    input_documents: relevantDocs,
    question: input,
  });

  return res.output_text;
};
