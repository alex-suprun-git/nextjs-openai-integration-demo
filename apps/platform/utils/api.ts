export const createUrl = (path: string) => window.location.origin + path;

export const updateUserPromptUsage = async (promptContentLength: number) => {
  const res = await fetch(
    new Request('/api/user', {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ promptContentLength }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const createNewEntry = async (content: string) => {
  const res = await fetch(
    new Request('/api/journal', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

// @Deprecated
export const updateEntry = async (id: string, content: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data.data;
};

export const deleteEntry = async (id: string) => {
  const res = await fetch(
    new Request(`/api/journal/${id}`, {
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' },
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  return;
};

export const askQuestion = async (question: string) => {
  const res = await fetch(
    new Request('/api/question', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ question }),
    }),
  );

  if (!res.ok) {
    return new Error(res.statusText);
  }

  const data = await res.json();
  return data;
};
