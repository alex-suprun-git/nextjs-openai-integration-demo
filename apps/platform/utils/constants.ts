export const MINIMUM_CONTENT_LENGTH = 30;
export const COOKIE_NAME = 'NEXT_LOCALE';
export const ENTRIES_BASE_PATH = '/records';
