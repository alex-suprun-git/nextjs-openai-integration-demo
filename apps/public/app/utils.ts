export const createUrl = (path: string) => window.location.origin + path;

export const getPageURL = (url: string) => {
  const [, , ...restSegments] = url.split('/');
  return restSegments.join('/');
};

export async function fetcher(_url: string) {
  const url = createUrl(_url);

  const res = await fetch(
    new Request(url, {
      method: 'GET',
    }),
  );

  if (!res.ok) return;
  return res.json() as Promise<{ isSignedIn: boolean }>;
}
