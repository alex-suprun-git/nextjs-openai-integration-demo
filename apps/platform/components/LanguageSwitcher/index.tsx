'use client';

import { MouseEvent } from 'react';
import { GrLanguage } from 'react-icons/gr';
import { setUserLocale } from '@/utils/locales';
import { useLocale } from 'next-intl';

const LanguageSwitcher = () => {
  const locale = useLocale();

  const languages = [
    { id: '01', label: 'English', value: 'en' },
    { id: '02', label: 'Deutsch', value: 'de' },
  ];

  const handleLanguageChange = (e: MouseEvent<HTMLAnchorElement>, language: UserLocale) => {
    e.preventDefault();
    setUserLocale(language);
    const elem = document.activeElement as HTMLElement;
    if (elem) {
      elem?.blur();
    }
  };

  return (
    <div className="dropdown dropdown-hover dropdown-left ml-auto">
      <div tabIndex={0} role="button" className="btn m-1 border-0 bg-slate-800 hover:bg-slate-800">
        <GrLanguage />
      </div>
      <ul
        tabIndex={0}
        className="dropdown-content menu z-[1] w-36 rounded-box bg-slate-800 p-2 shadow sm:w-40"
      >
        {languages.map((language) => {
          const isCurrentLang = language.value === locale;
          return (
            <li key={language.id}>
              <a
                href="#"
                className=""
                onClick={(e) => handleLanguageChange(e, language.value as UserLocale)}
              >
                {isCurrentLang ? <strong>{language.label}</strong> : language.label}
              </a>
            </li>
          );
        })}
      </ul>
    </div>
  );
};

export default LanguageSwitcher;
