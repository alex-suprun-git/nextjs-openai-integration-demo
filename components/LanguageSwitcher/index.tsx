'use client';

import { MouseEvent } from 'react';
import { GrLanguage } from 'react-icons/gr';
import { setUserLocale } from '@/utils/locales';

const LanguageSwitcher = () => {
  const languages = [
    { label: 'English', value: 'en' },
    { label: 'Deutsch', value: 'de' },
  ];

  const handleLanguageChange = (e: MouseEvent<HTMLAnchorElement>, language: UserLocale) => {
    e.preventDefault();
    setUserLocale(language);
    const elem = document.activeElement as HTMLElement;
    if (elem) {
      elem?.blur();
    }
  };

  return (
    <div className="dropdown dropdown-left dropdown-hover ml-auto">
      <div tabIndex={0} role="button" className="btn m-1 bg-slate-800 hover:bg-slate-800">
        <GrLanguage />
      </div>
      <ul
        tabIndex={0}
        className="menu dropdown-content z-[1] w-52 rounded-box bg-base-100 p-2 shadow"
      >
        {languages.map((language, index) => (
          <li key={index}>
            <a href="#" onClick={(e) => handleLanguageChange(e, language.value as UserLocale)}>
              {language.label}
            </a>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default LanguageSwitcher;
