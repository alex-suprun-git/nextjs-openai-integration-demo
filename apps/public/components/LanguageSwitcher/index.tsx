'use client';

import { usePathname } from 'next/navigation';
import { GrLanguage } from 'react-icons/gr';
import Link from 'next/link';
import { useLocale } from 'next-intl';
import { getPageURL } from '@/app/utils';

const LanguageSwitcher = () => {
  const pathname = usePathname();
  const locale = useLocale();
  const slug = getPageURL(pathname);

  const languages = [
    { id: 'en', label: 'English', value: `/en/${slug}` },
    { id: 'de', label: 'Deutsch', value: `/de/${slug}` },
  ];

  return (
    <div className="dropdown dropdown-hover dropdown-bottom ml-auto sm:dropdown-left">
      <div tabIndex={0} role="button" className="btn m-1 border-0 bg-slate-800 hover:bg-slate-800">
        <GrLanguage />
      </div>
      <ul
        tabIndex={0}
        className="dropdown-content menu z-[1] w-36 rounded-box bg-slate-800 p-2 shadow sm:w-40"
      >
        {languages.map((language) => {
          const isCurrentPage = language.value === `/${locale}/${slug}`;

          return (
            <li key={language.id}>
              <Link href={language.value} className="">
                {isCurrentPage ? <strong>{language.label}</strong> : language.label}
              </Link>
            </li>
          );
        })}
      </ul>
    </div>
  );
};

export default LanguageSwitcher;
