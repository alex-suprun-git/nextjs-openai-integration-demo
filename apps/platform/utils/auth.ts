import { getServerSession } from 'next-auth';
import { authOptions } from '@/app/api/auth/[...nextauth]/auth';
import { prisma } from './db';
import { Prisma } from '@prisma/client';
import { isDynamicServerError } from 'next/dist/client/components/hooks-server-context';

export const getCurrentUser = async () => {
  try {
    const session = await getServerSession(authOptions);

    if (!session?.user?.email) {
      console.warn('User email is not available. User might be signed out.');
      return null;
    }

    const user = await prisma.user.findFirstOrThrow({
      where: {
        email: session.user.email as string,
      },
    });

    return user;
  } catch (error: any) {
    if (error instanceof Prisma.PrismaClientValidationError) {
      console.error('Prisma validation error:', error.message);
      throw new Error('Failed to find user due to validation error.');
    } else if (error.code === 'P2025') {
      console.error('User not found:', error.message);
      throw new Error('User not found.');
    } else {
      if (isDynamicServerError(error)) {
        throw error;
      }
      console.error('Unexpected error:', error);
      throw new Error('An unexpected error occurred.');
    }
  }
};
