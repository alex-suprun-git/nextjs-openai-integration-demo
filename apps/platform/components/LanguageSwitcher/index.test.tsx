import { vi } from 'vitest';
import { screen, render, fireEvent } from '@testing-library/react';
import { setUserLocale } from '@/utils/locales';
import LanguageSwitcher from '.';

// Mock setUserLocale
vi.mock('@/utils/locales', () => ({
  setUserLocale: vi.fn(),
}));

describe('LanguageSwitcher Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    const { container } = render(<LanguageSwitcher />);
    expect(container).toMatchSnapshot();
  });

  it('calls setUserLocale with the correct language when a language is clicked', () => {
    render(<LanguageSwitcher />);

    const englishLink = screen.getByText('English');
    fireEvent.click(englishLink);
    expect(setUserLocale).toHaveBeenCalledWith('en');

    const germanLink = screen.getByText('Deutsch');
    fireEvent.click(germanLink);
    expect(setUserLocale).toHaveBeenCalledWith('de');
  });
});
