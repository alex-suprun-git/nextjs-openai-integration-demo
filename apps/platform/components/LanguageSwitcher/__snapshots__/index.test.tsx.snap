// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LanguageSwitcher Component > renders correctly 1`] = `
<div>
  <div
    class="dropdown dropdown-hover dropdown-left ml-auto"
  >
    <div
      class="btn m-1 border-0 bg-slate-800 text-stone-300 hover:bg-slate-800"
      role="button"
      tabindex="0"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12,23 C18.0751322,23 23,18.0751322 23,12 C23,5.92486775 18.0751322,1 12,1 C5.92486775,1 1,5.92486775 1,12 C1,18.0751322 5.92486775,23 12,23 Z M12,23 C15,23 16,18 16,12 C16,6 15,1 12,1 C9,1 8,6 8,12 C8,18 9,23 12,23 Z M2,16 L22,16 M2,8 L22,8"
          fill="none"
          stroke-width="2"
        />
      </svg>
    </div>
    <ul
      class="dropdown-content menu z-[1] w-36 rounded-box bg-slate-800 p-2 shadow sm:w-40"
      tabindex="0"
    >
      <li>
        <a
          class="text-stone-300"
          href="#"
        >
          English
        </a>
      </li>
      <li>
        <a
          class="text-stone-300"
          href="#"
        >
          Deutsch
        </a>
      </li>
    </ul>
  </div>
</div>
`;
